lib/marcel/mime_type.rb

```diff
@@ -5,10 +5,37 @@ class MimeType
     BINARY = "application/octet-stream"
 
     class << self
-      def extend(type, extensions: [], parents: [], magic: nil)
-        extensions = (Array(extensions) + Array(Marcel::TYPE_EXTS[type])).uniq
-        parents = (Array(parents) + Array(Marcel::TYPE_PARENTS[type])).uniq
-        Magic.add(type, extensions: extensions, magic: magic, parents: parents)
+      def canonicalize(type, instead_of:)
+        Magic.canonicalize type, instead_of: instead_of
+      end
+
+      def extend(type, extensions: nil, aliases: nil, parents: nil, magic: nil)
+        extensions = Array(extensions)
+        if extensions.any? && extensions.sort == Array(Marcel::TYPE_EXTS[type]).sort
+          warn "#{type} already has extensions #{extensions.inspect}"
+        end
+        extensions |= Array(Marcel::TYPE_EXTS[type])
+
+        aliases = Array(aliases)
+        existing_aliases = Marcel::TYPE_ALIASES.select { |_, t| t == type }.keys
+        if aliases.any? && aliases.sort == existing_aliases.sort
+          warn "#{type} already has aliases #{aliases.inspect}"
+        end
+        aliases |= existing_aliases
+
+        parents = Array(parents)
+        if parents.any? && parents.sort == Array(Marcel::TYPE_PARENTS[type]).sort
+          warn "#{type} already has parents #{parents.inspect}"
+        end
+        parents |= Array(Marcel::TYPE_PARENTS[type])
+
+        magic = Array(magic)
+        existing_magic = Marcel::MAGIC.select { |type, _| type == type }.map(&:last)
+        if magic.any? && magic == existing_magic
+          warn "#{type} already has magic matchers #{magic.inspect}"
+        end
```

Comment:
This is all pretty grotesque. Trying to limit the API footprint and not mess with the carefully tuned data tables, though.